Initial documentation of OpenTelemetry, splitting monitoring-and-metrics (#282)
AxonIQ/reference-guide: axon-framework/monitoring/tracing.md

# Distributed Tracing

Distributed Tracing enables you to track the path of a message through your system to see how the system behaves and
performs.
Axon Framework provides additional tracing functionality to track what takes time in your microservice, such as how long

> it took to load the aggregate, how long the actual command invocation took, or how long it took to publish events.

Reviewer: What do you think about listing the three options right away here, providing a link to those sections in this file as well.

Author: I think that's quite ugly. What I've done, is move the terminology section down. The table with all options should be immediately visible to the user now

> 

Author: Added a deprecation warning

## Terminology

A **trace** is a collection of one or more **spans** that together form a complete journey through your software.
Creating a span that is not part of a trace will automatically create one with that span being the root span of the
trace.

> Tools such as _ElasticSearch APM_ can render tracing information, as visible in the following image:

Author: Added a link to ES APM

| OpenTelemetry    | Yes       | [OpenTelemetry](https://opentelemetry.io/docs/concepts/what-is-opentelemetry/) is the successor of OpenTracing, with auto-instrumentation being its most prominent feature.   |
| OpenTracing      | Limited   | [OpenTracing](https://opentracing.io/) is supported [by an extension with limited functionality](../../extensions/tracing.md). Usage of OpenTelemetry is recommended instead. |
| SLF4j            | Yes       | If you have no monitoring system in place but want to trace through logging, the framework provides a `LoggingSpanFactory`.                                                   |

You configure a `SpanFactory` in the following ways:

{% tabs %}
{% tab title="Spring Boot Bean" %}

```java
@Configuration
public class AxonConfiguration {

    @Bean
    public SpanFactory spanFactory() {
        // Any bean implementing the SpanFactory will be picked up automatically and override the defaults
        return new MyCustomSpanFactory();
    }
}
```

{% endtab %}

> {% tab title="Axon Configurer" %}

Author: I have moved it, but I'm not sure I agree. Most people know Spring boot configuration, but not Axon configuration. I think we should show the most familiar thing first to improve comprehension

```java
public class AxonConfigurer {

    public void configure(Configurer configurer) {
        configurer.configureSpanFactory(configuration -> new MyCustomSpanFactory());
    }
}
```

Note that this is not necessary for all providers, since some may provide Spring Boot autoconfiguration out of the box.

> Refer to the specific subsections of this page on how to configure that specific `SpanFactory`.

Author: Most will configure only one. I changed the wording

{% endtab %}

### Combining factories

Sometimes you want the functionality of multiple `SpanFactory` implementations,
while Axon's configuration only allows one.
For this purpose, the framework contains the `MultiSpanFactory`,
which can be configured with multiple factories to which it delegates its calls.

For example, you can configure both the `LoggingSpanFactory` and the `OpenTelemetrySpanFactory` in the following
fashion:

{% tabs %}
{% tab title="Spring Boot Bean" %}

```java
@Configuration
public class AxonConfiguration {

    @Bean
    public SpanFactory spanFactory() {
        return new MultiSpanFactory(
                Arrays.asList(
                        LoggingSpanFactory.INSTANCE,
                        OpenTelemetrySpanFactory
                                .builder()
                                .build()
                )
        );
    }
}
```

{% endtab %}

> {% tab title="Axon Configurer" %}

Author: Again, not sure I agree but I applied it.

the name, and a message that triggered the span (if any, it's not required). The framework can request
the span types defined in the following table:

| Span Type     | Description                                                                                      | 
|---------------|--------------------------------------------------------------------------------------------------|
| Root trace    | Create a new trace entirely, having no parent.                                                   |
| Dispatch span | A span which is dispatching a message.                                                           |
| Handler span  | A span which is handling a message. Will set the span that dispatched the message as the parent. |
| Internal span | A span which specified something internal. It's not an entry or exit point.                      |

A trace generally consists of multiple spans with different types, depending on the functionality.
When a message is dispatched and handled by another process or thread, the handling trace is related to the dispatching
trace.
This handling trace becomes part of the root trace unless it's an asynchronous invocation. In that case, it will be

> linked to it instead,

Author: Added an explanation :)

so tooling can refer to the related traces but the traces don't become so large they are unreadable.

The following functionality in Axon Framework is traced in addition to the tracing capabilities already provided by the
standard of your choice:

- Dispatching and handling of commands
- Publishing of events
- Handling of events by event processors
- Dispatching and handling of queries
- Scheduling and invocation of deadlines
- Creation of snapshots

> - Loading of the aggregate, including time spent waiting for a lock

Reviewer: How about we link to the chapters of each in this list?
So, we link to Command dispatching/handling, Event publication, Event Processors, Query dispatching/handling, deadlines, snapshots, [repositories](https://docs.axoniq.io/reference-guide/axon-framework/axon-framework-commands/configuration#command-model-repositories), and sagas.

Author: Agreed, adding this

- Axon Server GRPC calls

> - Tracing of each message handler invocation (Spring Boot only)

Reviewer: Doesn't this include sagas too?

and add it to the `SpanFactory`, in case you want to add custom information you would like to include as a label.

```java
public class CustomSpanAttributesProvider implements SpanAttributesProvider {

    @Nonnull
    @Override
    public Map<String, String> provideForMessage(@Nonnull Message<?> message) {
        // Provide your labels based on the message here
        return Collections.emptyMap();
    }
}
```

You can register this custom `SpanAttributesProvider` in one of the following ways.

{% tabs %}
{% tab title="Spring Boot Bean" %}

```java
@Configuration
public class AxonConfiguration {

    @Bean
    public SpanAttributesProvider customSpanAttributesProvider() {
        // Autoconfiguration picks beans of type SpanAttributesProvider up automatically.
        return new CustomSpanAttributesProvider();
    }
}
```

{% endtab %}
{% tab title="Spring Boot injection" %}

```java
@Configuration
public class AxonConfiguration {

    @Autowired
    public void configureSpanFactory(SpanFactory spanFactory) {
        spanFactory.registerSpanAttributeProvider(new CustomSpanAttributesProvider());
    }
}
```

{% endtab %}  

> {% tab title="Manual" %}

Reviewer: Again I'd prefer the plain Axon configuration before any Axon Framework enhancements (like Spring).

```java
public class AxonConfigurer {

    public void configure(Configuration configuration) {

>         configuration.spanFactory().registerSpanAttributeProvider(new CustomSpanAttributesProvider());

Reviewer: Would this really work? I'd honestly expect the configuration of a custom `SpanAttributesProvider` to occur on a `Configurer` / `ConfigurerModule`.

Author: Yes, it's a contract of the `SpanFactory` interface that attribute providers can be added later. For autoconfiguration we do this with a `ConfigurerModule`, but the user can do that later.

    }
}
```

{% endtab %}
{% endtabs %}

## OpenTelemetry <a id="opentelemetry"></a>

Axon work provides [OpenTelemetry support](https://opentelemetry.io/docs/concepts/what-is-opentelemetry/) out of the

> box. The OpenTelemetry standard improves upon the OpenTracing standard by providing more

Reviewer: ```suggestion
box. The OpenTelemetry standard improves upon the OpenTracing and OpenCensus standard by providing more
```

auto-instrumentation without the need for the user to configure many things.

OpenTelemetry works by adding a Java agent to the execution of the application. Based on the configuration, the agent
will collect logs, metrics and tracing automatically before sending it to a collector that can provide insights.

> ElasticSearch APM, Jaeger and many other tools are available for this and the configuration of these is beyond the scope

Reviewer: I assume that "this" references to collecting insights?

Author: Adjusted wording here

of this guide. You can find more
information [in the "Getting Started" section of the OpenTelemetry documentation.](https://opentelemetry.io/docs/instrumentation/java/getting-started/)

OpenTelemetry [supports a lot of libraries,
frameworks and application servers out of the box.](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/supported-libraries.md)
For example, when a Spring REST endpoint is called it will automatically start a trace.
With the `axon-tracing-opentelemetry` module, this trace will be propagated to all subsequent Axon Framework messages.
For example, if the REST call produces a command which is sent over Axon Server,
handling the command will be included in the same trace as the original REST call.

### Configuration

To get OpenTelemetry support enabled you will need to add the following dependency to your application's dependencies:

{% tabs %}

> {% tab title="Maven" %}

Reviewer: This would only be sufficient with Spring Boot, right?
Without it, you will need to add the dependency *and* configure the `SpanFactory` to be an `OpenTelemetrySpanFactory`.

Author: Yes, that's in the next sentence: `Depending on your application, more configuration might be needed.`
There are two subsections, one for boot and one for custom

```xml  
<dependency>  
    <groupId>org.axonframework</groupId>  
    <artifactId>axon-tracing-opentelemetry</artifactId>
	<version>${axon-framework.version}</version>
</dependency>
```

{% endtab %}%}
{% tab title="Gradle" %}

```gradle
implementation group: 'org.axonframework', name: 'axon-tracing-opentelemetry', version: axonFrameworkVersion
```

{% endtab %}
{% endtabs %}

Depending on your application, more configuration might be needed.

> ## Spring boot autoconfiguration

Reviewer: I'd wager this is a subsection of the OpenTelemetry chapter? If so, I think you need to add one (#) or two (##) hashes to make it align as intended.

> ## Spring boot autoconfiguration

Reviewer: ```suggestion
## Spring Boot auto configuration
```

> When using the Spring Boot autoconfiguration of Axon Framework, most things will be autoconfigured regardless of the

Reviewer: I may be wrong, but I think we should split "auto" and "configuration."
Or, perhaps we add a hyphen? Any how, I believe it's not one word.
```suggestion
When using the Spring Boot auto configuration of Axon Framework, most things will be auto configured regardless of the
```

| `axon.tracing.attributeProviders.aggregateIdentifier` | `true`  | Whether to add the aggregate identifier as a label when handling a message                             |
| `axon.tracing.attributeProviders.messageId`           | `true`  | Whether to add the message identifier as a label when handling a message                               |
| `axon.tracing.attributeProviders.messageName`         | `true`  | Whether to add the message name as a label when handling a message                                     |
| `axon.tracing.attributeProviders.messageType`         | `true`  | Whether to add the message type as a label when handling a message                                     |
| `axon.tracing.attributeProviders.payloadType`         | `true`  | Whether to add the payload type as a label when handling a message                                     |
| `axon.tracing.attributeProviders.metadata`            | `true`  | Whether to add the metadata properties as labels when handling a message                               |

> #### Manual configuration

Reviewer: Ah, here it is. I'd expect this to be part of a code tab I guess. Any how, not after all the Spring Boot configuration stuff.

Author: I'm not sure what you mean. It's in the auto-configuration section of OpenTelemetry.  Where else would you like it?

The OpenTelemetry support can also be configured using the `Configurer` of Axon Framework to configure
the `OpenTelemetrySpanFactory`.

```java
public class AxonConfigurer {

    public void configure(Configurer configurer) {
        configurer.defaultConfiguration()
                  .configureSpanFactory(c -> OpenTelemetrySpanFactory.builder().build());
    }
}
```

Note that when not using Spring boot, tracing each message handler invocation is not supported due to a limitation.

## OpenTracing <a id="opentracing"></a>

> OpenTracing is deprecated. If necessary, you can
> use [the OpenTracing extension to Axon Framework](../../extensions/tracing.md).

Reviewer: I'd expand a little here for clarity:
```suggestion
Although available, know that the OpenTracing standard is deprecated. 
If necessary, you can still use [the OpenTracing extension of Axon Framework](../../extensions/tracing.md), though.

```

> The functionality of the extension is rather limited compared to the OpenTelemetry integration, so using that if
> possible is preferred.

Reviewer: ```suggestion
Note that the functionality of this extension is limited compared to the OpenTelemetry integration.
Hence, if possible, it is recommended to switch to OpenTelemetry.
```

## Logging <a id="logging"></a>

Sometimes you don't have an APM system available, for instance during local development. It might still be useful to see
the traces that would be started and finished to obtain insights. For this purpose, the framework provides
a `LoggingSpanFactory`.

You can configure the `LoggingSpanFactory` in the following ways:
{% tabs %}
{% tab title="Spring Boot Bean" %}

```java
@Configuration
public class AxonConfiguration {

    @Bean
    public SpanFactory spanFactory() {
        return LoggingSpanFactory.INSTANCE;
    }
}
```

{% endtab %}

> {% tab title="Manual" %}

Reviewer: Again I'd prefer the plain Axon configuration before any Axon Framework enhancements (like Spring).

Reviewer: Added, I think it's good to keep the titles of these consistent among the code blocks.

```java
public class AxonConfigurer {

    public void configure(Configuration configuration) {
        configuration.configureSpanFactory(c -> LoggingSpanFactory.INSTANCE);
    }
}
```

{% endtab %}
{% endtabs %}

## Traced components

Axon Framework provides a large range of components that are traced by the configured `SpanFactory`.

> Each component and their added spans are available for reference in this section, with additional information about how

Reviewer: ```suggestion
Each component, and their added spans, are available for reference in this section, with additional information about how
```

Author: Changed wording, but I don't like the two comma approach. Grammarly says both are valid

they should be interpreted.

It's important to note that the availability of these spans is highly dependent on the application configuration. 
For instance, some components are only used when using Axon Server, or you might have created your own `CommandBus` 
implementation which does not call the `SpanFactory` API. 

### Commands

The `CommandBus` is instrumented to create spans for both dispatching and handling commands. 

> When using the `AxonServerCommandBus`, there will be two handling and dispatch traces 

Reviewer: This makes me wonder: how would this act with the `DistributedCommandBus`?
It's similar to Axon Server here, as it also uses a `localSegment` (guess who was first, though).
Not something to solve right away, but *definitely* something to look into for 4.7.0.

Reviewer: Similarly...have you checked what the behavior is with the `DisruptorCommandBus`? I think we missed that one too :see_no_evil: 
Differently put, something else to look into.

Author: Yes, let's put that on the 4.7.0 list :)

since it uses a second `CommandBus` to invoke the command locally after receiving it from Axon Server. In addition,
you can see the GRPC-call to Axon Server and the time it took to handle the call.

The traces will, in sequence, look like this:

| Trace name                                      | Description                                                                                                                                                                    |

> |-------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

Reviewer: Just a though, but how about we use those code blocks to provide the different scenarios?
So, one with AxonServer, one without, one with the `DistributedCommandBus` (in the future), etc.

Author: I do like the idea and will apply it. I am, however, very afraid this will blow up to a lof of tabs

| `${CommandBusClass}.dispatch(${commandName})`   | The bus is dispatching the command locally (If using Axon Server: this is invoked by the `AxonServerCommandBus`.                                                               |
| `${CommandBusClass}.handle(${commandName})`     | The bus is invoking the handler locally.                                                                                                                                       |
| `${RepositoryClass}.load ${identifier}`         | The aggregate is being loaded by the repository. During this time Axon Framework will obtain a lock, fetch snapshots and events from the event store to hydrate the aggregate. |
| `LockingRepository.obtainLock`                  | The repository is obtaining a lock for the aggregate. This taking some time indicates that the command was queued due to another command being handled for the same aggregate. |

During handling of commands, other functionality might be invoked such as scheduling deadlines or publishing events. 
Please refer to the specific sections of this functionality for more information.

### Events

> Publishing events during command execution happens in two steps:

Reviewer: Although I know the chance is slim, event publication is not restricted to command handling only.  Users can invoke the `EventGateway` too.
I'd expect this section to provide that scenario, and trace sample, too.

| `${EventBusClass}.publish(${EventClass})` | For each event, a short span is created to indicate that an event was published. |
| `${EventBusClass}.commit`                 | Indicates events being committed to the event store.                             |

The events have their own specific publishing span to link other spans to it that are related.
Any streaming event processor or saga handling the event in the future will be linked
to publishing spans, allowing easy click-through.

### Event processors

Event processor invocations are traced as well. They will add the following traces:

| Trace name                                                  | Description                                                                 |
|-------------------------------------------------------------|-----------------------------------------------------------------------------|
| `${ProcessorType}[${processorName}](${EventClass})`         | Root trace of handling the event, includes all interceptor invocations.     |
| `${ProcessorType}[${processorName}].process(${EventClass})` | Inner span of handling the event, after all interceptors have been invoked. |

> In case of a `SubscribingEventProcessor`, the invocation happens during command execution.

Reviewer: This is also not restricted to command execution, due to the AMQP and Kafka Extension. Please provide those options here too.

No new root trace will be created, but the inner process span of the processor is added to indicate the invocation.

### Deadlines

Any action related to deadlines is traced in order to gain insight into what happened during specific calls. 
Mutations on deadlines generally happen from another root trace, such as a command or saga.
The handling span of a deadline will be linked to the scheduling span for easy navigation.

| Trace name                                                               | Description                                                                |
|--------------------------------------------------------------------------|----------------------------------------------------------------------------|
| `${DeadlineManagerClass}.schedule(${deadlineName})`                      | A deadline was scheduled.                                                  |

> | `${DeadlineManagerClass}.cancelSchedule(${deadlineName}, ${scheduleId})` | A deadline was cancelled based on name and scheduleId.                     |

Reviewer: ```suggestion
| `${DeadlineManagerClass}.cancelSchedule(${deadlineName}, ${scheduleId})` | A deadline was cancelled based on name and `scheduleId`.                     |
```

However, it can still be useful to measure the performance of snapshotting and see when it is triggered.
The root trace of the `Snapshotter` invocation will be linked to the command handling span 
after which the snapshot was scheduled to be created.

| Trace name                                                                  | Description                                                                                                           |
|-----------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| `${SnapshotterClass}.createSnapshot($aggregateClass)`                       | A snapshot creation task is being submitted. Depending on performance, the executor might take a while to pick it up. |
| `${SnapshotterClass}.createSnapshot($aggregateClass, $aggregateIdentifier)` | The `Snapshotter` is now creating the snapshot.                                                                       |

Note that the root trace does not contain the aggregate identifier,

> so any snapshot invocation of the same aggregate type is grouped together by the APM tool.

Reviewer: Nit: I'd make a separate sentence out of "so any snapshot..." for clarity..

Author: I changed the sentence, I kind of like it in the new form. It feels wrong to split up

> ### Sagas

Reviewer: Why have you stopped adding an empty line after the chapter? :cry: 
```suggestion
### Sagas

```

Sagas are a special type of event processor that can invoke multiple saga's for a single event. 
Because of this the `AbstractSagaManager` has been instructed with additional tracing information.
These spans are descendants of an event processor span that invokes the manager.

| Trace name                                                                  | Description                                          |
|-----------------------------------------------------------------------------|------------------------------------------------------|
| `SagaManager[${SagaTypeName}].invokeSaga ${sagaIdentifier}`                 | A matching saga has been found and is being invoked. |
| `SagaManager[${SagaTypeName}].startNewSaga`                                 | The manager is constructing a new saga.              |

> ### Queries

Reviewer: ```suggestion
### Queries

```

Queries support tracing in all of their forms. In order to be clear about how they work, 
this section is split based upon the query's type. 
For all types, the created spans will differ based on whether Axon Server is used or not. 
The spans that are only available with Axon Server are marked as such.

> #### Direct queries

Reviewer: ```suggestion
#### Direct queries

```

Direct queries fetch a single result (either a single item or a single list) and receive no updates. 
Their traces will look like the following table:

> | Trace name                                                | Description                                                                           |

Reviewer: I'd have a similar code-block suggestion for the AxonServer and non-AxonServer samples, just as with the Commands section.

|-----------------------------------------------------------|---------------------------------------------------------------------------------------|
| `AxonServerQueryBus.query(${queryName})`                  | (Only with Axon Server) The requesting service is dispatching the query.              |
| `QueryProcessingTask(${queryName})`                       | (Only with Axon Server) The handling service is handling the query request in a task. |
| `SimpleQueryBus.query(${queryName})`                      | The handling service is handling the query.                                           |
| `AxonServerQueryBus.ResponseProcessingTask(${queryName})` | (Only with Axon Server) The requesting service is processing the response.            |

#### Streaming queries

> Streaming queries, introduced in Axon Framework 4.6.0, look very similar to the traces of a Direct query. 

Reviewer: The selected reference guide version will automatically state it's Axon 4.6.0.
Hence, there's no need to specify that again.

```suggestion

Streaming queries look very similar to the traces of a Direct query. 
```

They do not contain a `ResponseProcessingTask` span since their results are directly published to the invoker of the query.
Their traces will look like the following table:

> | Trace name                                        | Description                                                                           |

Reviewer: I'd have a similar code-block suggestion for the AxonServer and non-AxonServer samples, just as with the Commands section.

|---------------------------------------------------|---------------------------------------------------------------------------------------|
| `AxonServerQueryBus.streamingQuery(${queryName})` | (Only with Axon Server) The requesting service is dispatching the query.              |
| `QueryProcessingTask(${queryName})`               | (Only with Axon Server) The handling service is handling the query request in a task. |
| `SimpleQueryBus.query(${queryName})`              | The handling service is handling the query.                                           |

> #### Scatter-Gather queries

Reviewer: ```suggestion
#### Scatter-Gather queries

```

Scatter-Gather queries are like a direct query but can fetch results from multiple services at the same time. 
Part of the trace can thus be duplicated multiple times, since multiple services are invoked.
Their traces will look like the following table:

> | Trace name                                             | Description                                                                                          |

Reviewer: I'd have a similar code-block suggestion for the AxonServer and non-AxonServer samples, just as with the Commands section.

|--------------------------------------------------------|------------------------------------------------------------------------------------------------------|
| `AxonServerQueryBus.scatterGather(${queryName})`       | (Only with Axon Server) The requesting service is dispatching the query.                             |
| `QueryProcessingTask(${queryName})`                    | (Only with Axon Server) The handling service is handling the query request in a task.                |
| `SimpleQueryBus.scatterGather(${index})(${queryName})` | Each handling service is handling the query. Each handler within the same service has its own index. |

> #### Subscription queries

Reviewer: ```suggestion
#### Subscription queries

```

Subscription queries are traces in a different way than others. Subscription queries have an initial result, which is traces like a direct query.
However, new results can later be published at any time after while the caller is still subscribed to it. 

In order to prevent malformed traces, since most APM tools have a maximum span time before flushing them, publication of
new results is not part of the original trace. 
However, invocations of the `SimpleQueryUpdateEmitter` will be linked to
the span of the queries that are listening to it, so the original call can easily be found.

> The `UpdateEmitter` traces will look like the following table:

Reviewer: ```suggestion
The `QueryUpdateEmitter` traces will look like the following table:
```

| Trace name                                                     | Description                                                                                          |
|----------------------------------------------------------------|------------------------------------------------------------------------------------------------------|
| `SimpleQueryUpdateEmitter.emit(${PayloadClass})`               | A new update is emitted.                                                                             |
| `SimpleQueryUpdateEmitter.emit ${queryName} (${PayloadClass})` | A new update is emitted for a specific consumer.                                                     |

In addition to this, the spans of the [direct queries section](#direct-queries) apply as well. 

### Message handler invocations

The `TracingHandlerEnhancerDefinition` automatically creates a span for each message handler invocation within your

> application. This is true for commands, events, queries and event custom message handlers. Spans will be created with the following format:

Reviewer: ```suggestion
application. This is true for commands, events, queries and even custom message handlers. Spans will be created with the following format:
```
